### ISSUE TITLE
Clob.getCharacterStream() behavior difference when called before vs after ResultSet is closed for varchar(max)

### ISSUE THREAD

**Issue by jonyschak** [Microsoft/mssql-jdbc#800]
## Driver version
6.5.2+

## SQL Server version
Microsoft SQL Server 2017

## Client Operating System
Windows 7 x64

## JAVA/JVM version
JDK8u121 x64

## Table Schema
create table Test (
id numeric(18,0) not null,
data varchar(max)
);
[prepareTestData.txt](https://github.com/Microsoft/mssql-jdbc/files/2359556/prepareTestData.txt)

## Problem description
The fix for [#673](https://github.com/Microsoft/mssql-jdbc/issues/673) causes getCharacterStream() to behave differently if called before vs after the ResultSet is closed.

When you call it before the ResultSet is closed, getCharacterStream() is hardcoded to use UTF_16LE as the encoding.  If you close the ResultSet first, the closing of the ResultSet causes fillFromStream() to be called which calls getStringFromStream() which uses the charset from typeInfo to create the string, which is not UTF_16LE if the column is a varchar(max).

## Reproduction code
```
import java.sql.Statement;
import java.io.Reader;
import java.sql.Clob;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;

public class Test {
    private final static String CONNECTION_STRING = "jdbc:sqlserver://<>";
    private final static String USER = "<>";
    private final static String PWD = "<>";

    public static void main(String[] args) throws Exception {
        Connection connection = null;
        Statement statement1 = null;
        Statement statement2 = null;
        ResultSet rs1 = null;
        ResultSet rs2 = null;
        try {
            connection = DriverManager.getConnection(CONNECTION_STRING, USER, PWD);
            connection.setAutoCommit(false);

            // Select the desired data. 
            // See the attachment how to create the table and insert the data
            String selectSql = "SELECT data from Test where id= 1";
            statement1 = connection.createStatement();
            statement2 = connection.createStatement();
            rs1 = statement1.executeQuery(selectSql);
            rs1.next();
            rs2 = statement2.executeQuery(selectSql);
            rs2.next();

            // Access the clob column
            Clob c1 = rs1.getClob("data");
            System.out.println("Get clob1 column");
            Clob c2 = rs2.getClob("data");
            System.out.println("Get clob2 column");

            // Attempt to read data
            Reader is1 = c1.getCharacterStream();
            System.out.println("Reading data1: " + is1.read()); 

            // Free the result set
            rs1.close();
            System.out.println("Rs1 closed");
            rs2.close();
            System.out.println("Rs2 closed");

            statement1.close();
            System.out.println("Statement1 closed");
            statement2.close();
            System.out.println("Statement2 closed");

            // Attempt to read data
            Reader is2 = c2.getCharacterStream();
            System.out.println("Reading data2: " + is2.read()); 

        } finally {
            if (rs1 != null) {
                rs1.close();
            }
            if (rs2 != null) {
                rs2.close();
            }
            if (statement1 != null) {
                statement1.close();
            }
            if (statement2 != null) {
                statement2.close();
            }
            if (connection != null) {
                connection.close();
            }
        }
    }
}
```

Output from the sample provided

```
Get clob1 column
Get clob2 column
Reading data1: 28499
Rs1 closed
Rs2 closed
Statement1 closed
Statement2 closed
Reading data2: 83
```

**Comment by rene-ye** [comment]
Hi @jonyschak, as you mentioned, this issue is being addressed in #799. If you would like to take advantage of caching strings, moving the result set pointer or closing the result set without closing the Clob/NClob object will store the LoB in memory.

**Comment by rene-ye** [comment]
#799 has been merged and will be available starting from the next preview release. Closing the issue.
